### ISSUE TITLE
Not-deterministic errors on Python 3 for graph example

### ISSUE THREAD

**Issue by matthew-brett** [BIDS/datarray#75]
We are getting failures most of the time for the tests on `examples/inference_algs.py` - e.g. https://travis-ci.org/BIDS/datarray/jobs/136040496#L369 .

The test same test always passes on Python 2.

On Python 3, about 4 in 5 runs of `python examples/inference_algs.py` generates an error, with a few different wrong answers.

The problem is that:

```
margs3, lik3 = calc_marginals_sumproduct(cpts,evidence)
```

generates different results on different runs.   Specifically, this call always generates the same value for `lik3` within a single Python session, but a different value in different sessions.

Playing with the script, I suspect this is because the graph the script is operating on has a different sequence in different runs, and I believe the depth-first search will differ in that case.   I suspect this is a dictionary ordering problem, but I haven't tracked it down.

@terhorst - I think this is your code - do you have any idea what is going on?

**Comment by terhorst** [comment]
No idea, sorry. (I can barely remember working on this).

**Comment by matthew-brett** [comment]
Ah - well - the algorithm does seem to depend on dictionary ordering.  I think this is because networkx uses the dictionary node ordering for the `dfs_tree` algorithm when the start nodes are not specified, and the output differs with different ordering.  So I believe this is a bug, in that it was coincidence that the code worked for Python 2.7, and the node order needs to be specified on the call to `dfs_tree` [here](https://github.com/BIDS/datarray/blob/master/examples/inference_algs.py#L358).

And in fact I can reproduce the errors on Python 2.7 with something like `PYTHONHASHSEED=42 python examples/inference_algs.py`.  

Unfortunately I don't understand the algorithm, so I don't know what the right nodes are for the dfs_tree search.

**Comment by matthew-brett** [comment]
Exploring further - I don't think it is the depth search, I think it is simply that [this block](https://github.com/BIDS/datarray/blob/d188dc17c06200fc6c0182e0336f78116044a396/examples/inference_algs.py#L308):

```
    for node in G.nodes():
        potential = multiply_potentials(*[messages[(src,node)] for src in G.neighbors(node)])
        marginals[node] = normalize(potential)
        
    return marginals, potential.sum()
```

is iterating over the elements of an underlying dictionary (`G.nodes()`), so `potential.sum()` at the end returns some value effectively randomly chosen from all the nodes in the dictionary.  I think that we actually want:

```
    for node in G.nodes():
        potential = multiply_potentials(*[messages[(src,node)] for src in G.neighbors(node)])
        marginals[node] = normalize(potential)
        potentials[node] = potential
        
    return marginals, potentials[target_node].sum()
```

where `target_node = 'burglary'`.  But I'm happy to be corrected.

**Comment by matthew-brett** [comment]
PR to fix at #76

**Comment by matthew-brett** [comment]
Closed by #76.
